### PR TITLE
Omnibus

### PR DESCRIPTION
**Pull request by boxerab**
# Omnibus 

This branch is going to pull in all of my pending pull requests into one single branch.
It adds performance improvements and better memory management.
Also, memory mapped and memory buffered streams are supported.

# Measurements

I compared master and omnibus branches for memory usage and performance.
(my system is a 4 year old 4-core i7 3770 with DDR3, running Windows 7 64)

##  Peak Memory Usage

For decode of 100 DCI frames:

1.  Master :    98 MB
2. Omnibus   35 MB 

So, memory usage has dropped to almost one third. 
 

## Performance:

Omnibus uses OpenMP to multi-thread the library.

For encode/decode of 100 DCI frames:

1. Master (single threaded)

Encode:  0.85 FPS
Decode : 2.5 FPS

2. Omnibus (multi threaded)

Encode:  5.5 FPS
Decode : 11.2 FPS

On upcoming 8 core FinFET desktop CPUs, this should get real time decode, and 
~15 FPS encode.

### CODE REVIEW

**Comment by XenonofArcticus** [comment]
This is fantastic work and I look forward to being able to use it.

**Comment by boxerab** [comment]
Thanks, Xenon !

**Comment by homme** [comment]
This looks great - many thanks! I'm particularly interested in using these improvements in conjunction with [GDAL](http://www.gdal.org/frmt_jp2openjpeg.html).

**Comment by boxerab** [comment]
Glad to be of service, homme. I've got a few more tricks up my sleeve, will be checking them in shortly.

**Comment by klokan** [comment]
This looks like cool performance improvement of OpenJPEG for the first glimpse!

**Comment by boxerab** [comment]
Thanks, Petr. 8 core CPUs are coming to the desktop this year; we need to make use of them.

**Comment by mayeut** [comment]
@boxerab Seems nice !
Could you edit your post to include performances before the patch on your system ? That would give a better idea of the improvement. (The same could be done for memory consumption)

BTW, there are many warnings on the build: http://my.cdash.org/viewBuildError.php?type=1&buildid=887331 Do you think you can get this down a bit ?

**Comment by boxerab** [comment]
Thanks @mayeut I will update perf and memory consumption stats from my system.

As for the warnings, yes, there are quite a lot. Will look into getting rid of them.

**Comment by boxerab** [comment]
By the way, my last commit on this branch made a small change to opj_image_compt_t, so it looks like I have changed the ABI. I need to do this because in certain cases, I am passing the tile data back to inside the user image, and this tile data is allocated using opj_aligned_malloc, so it must
be freed with opj_aligned_free.  This change is one of the keys to getting the memory down for certain work flows i.e. DCI decoding.  

Currently, peak memory usage for DCI decoding on my branch is around 1/3 of master (!) so it is now much more reasonable. But, the price we pay is we need to track whether image data was allocated with opj_aligned_malloc.

**Comment by boxerab** [comment]
Also, @mayeut , thanks again for setting up Travis and Appveyor. It made working out the problems with the branch pretty easy, although time consuming. It helped me a lot.

**Comment by mayeut** [comment]
@boxerab, even though the API/ABI check passes, some new declarations in `openjpeg.h` make me think that behavior has changed. There seems to be a need to call at least some of those new functions even when using only "old" API.

**Comment by boxerab** [comment]
Do you mean these methods?


    OPJ_API void OPJ_CALLCONV opj_image_all_components_data_free(opj_image_t* image);

    OPJ_API void OPJ_CALLCONV opj_image_single_component_data_free(opj_image_comp_t* image);

Yes, I suppose you are right. The user needs to call these when freeing up image data, they can't just call  `free(image->comps[0].data)`, for example. Or the program will crash.

If this is a problem, I can make a temporary change so these methods are not required. We will lose around 10% performance gain on decode, and add 1/3 memory usage back.

File: src/bin/common/color.c [hunk 1]
```diff
@@ -332,227 +384,234 @@ void color_apply_icc_profile(opj_image_t *image)
...
-	image->comps[1] = image->comps[0];
-	image->comps[2] = image->comps[0];
+		image->comps[1].data = new_image->comps[0].data;
+		image->comps[2].data = new_image->comps[1].data;
 
-	image->comps[1].data = (int*)calloc((size_t)max, sizeof(int));
-	image->comps[2].data = (int*)calloc((size_t)max, sizeof(int));
+		new_image->comps[0].data= NULL;
+		new_image->comps[1].data = NULL;
+
```
Comment (stweil) [inline src/bin/common/color.c]: Git reports a whitespace warning here (spaces at end of line).

File: src/bin/jp2/opj_decompress.c [hunk 2]
```diff
@@ -1235,33 +1238,89 @@ int main(int argc, char **argv)
...
 			}
 		}
+		*/
+		if (!l_stream) {
+			l_stream = opj_stream_create_mapped_file_read_stream(parameters.infile);
+		}
 
-		/* read the input file and put it in memory */
-		/* ---------------------------------------- */
+
```
Comment (stweil) [inline src/bin/jp2/opj_decompress.c]: Git reports spaces at end of line here.

File: src/lib/openjp2/CMakeLists.txt [hunk 3]
```diff
@@ -37,8 +47,12 @@ set(OPENJPEG_SRCS
   ${CMAKE_CURRENT_SOURCE_DIR}/pi.h
   ${CMAKE_CURRENT_SOURCE_DIR}/raw.c
   ${CMAKE_CURRENT_SOURCE_DIR}/raw.h
-  ${CMAKE_CURRENT_SOURCE_DIR}/t1.c
+  ${CMAKE_CURRENT_SOURCE_DIR}/segmented_stream.c
+  ${CMAKE_CURRENT_SOURCE_DIR}/segmented_stream.h
+  ${CMAKE_CURRENT_SOURCE_DIR}/t1.c
```
Comment (stweil) [inline src/lib/openjp2/CMakeLists.txt]: Spaces at end of line here and at the next but one line.

File: src/lib/openjp2/j2k.c [hunk 4]
```diff
@@ -8062,26 +8064,63 @@ OPJ_BOOL opj_j2k_decode_tile (  opj_j2k_t * p_j2k,
...
                 opj_j2k_tcp_destroy(l_tcp);
                 p_j2k->m_specific_param.m_decoder.m_state |= 0x8000;/*FIXME J2K_DEC_STATE_ERR;*/
                 opj_event_msg(p_manager, EVT_ERROR, "Failed to decode.\n");
                 return OPJ_FALSE;
         }
 
-        if (! opj_tcd_update_tile_data(p_j2k->m_tcd,p_data,p_data_size)) {
-                return OPJ_FALSE;
-        }
+
```
Comment (stweil) [inline src/lib/openjp2/j2k.c]: Spaces at end of line.

File: src/lib/openjp2/j2k.c [hunk 5]
```diff
@@ -8062,26 +8064,63 @@ OPJ_BOOL opj_j2k_decode_tile (  opj_j2k_t * p_j2k,
...
+				tilec->data = NULL;
+				comp->resno_decoded = p_j2k->m_tcd->image->comps[compno].resno_decoded;
+
+				/* now sanitize data (signed data is broken at the moment */
+				l_size_comp = (comp->prec + 7) >> 3;
+				if (l_size_comp <= 2) {
+					for (j = 0; j < comp->h; ++j)
+						for (i = 0; i < comp->w; ++i) {
+							if (l_size_comp == 1)
+								comp->data[i + j*comp->w] =
```
Comment (stweil) [inline src/lib/openjp2/j2k.c]: Space at end of line here, in the next line and also in lines 8119 and 8123.

File: src/lib/openjp2/j2k.c [hunk 6]
```diff
@@ -8165,14 +8210,9 @@ static OPJ_BOOL opj_j2k_update_image_data (opj_tcd_t * p_tcd, OPJ_BYTE * p_data,
 
                 /*-----*/
                 /* Compute the precision of the output buffer */
-                l_size_comp = l_img_comp_src->prec >> 3; /*(/ 8)*/
-                l_remaining = l_img_comp_src->prec & 7;  /* (%8) */
+                l_size_comp = (l_img_comp_src->prec + 7) >> 3;
```
Comment (stweil) [inline src/lib/openjp2/j2k.c]: Space at end of line.

File: src/lib/openjp2/j2k.c [hunk 7]
```diff
@@ -9545,24 +9585,79 @@ static OPJ_BOOL opj_j2k_allocate_tile_element_cstr_index(opj_j2k_t *p_j2k)
         return OPJ_TRUE;
 }
 
+static OPJ_BOOL opj_j2k_needs_copy_tile_data(opj_j2k_t *p_j2k) {
+	/* single tile, RGB images only*/
+	OPJ_BOOL copy_tile_data = (p_j2k->m_cp.th * p_j2k->m_cp.tw > 1) ||
+								p_j2k->m_output_image->numcomps != 3 ||
+								p_j2k->m_output_image->numcomps != 3 ||
+								p_j2k->m_tcd->tcd_image->tiles->numcomps != 3;
+
```
Comment (stweil) [inline src/lib/openjp2/j2k.c]: Spaces.

**Comment by boxerab** [comment]
@stweil  Thanks. 

I would like to get this merged in, and then we can worry about whitespace.
Or, you could make a patch to fix this for my omnibus branch, and then we can merge it in with whitespace fixed.

**Comment by boxerab** [comment]
@mayeut I have temporarily reversed some of my memory optimizations; now a user doesn't need to worry about allocating and de-allocating image memory using the library. So, nothing should break if a user upgrades to the upcoming release; no behaviour changes.

 I will put these optimizations back in for the following release.

**Comment by detonin** [comment]
Hi Aaron, thanks for this great work ! I really would like to merge it before the upcoming release but I have a few comments / questions:
* I see that although your changes do not break ABI, you added some functions to the API. This means we would go for version 2.2 rather than 2.1.1, according to http://semver.org/. This is ok for me, unless anyone has a comment on this.
* You added 4 new files to the library. We had in the past an issue concerning the OpenJPEG copyright which is a complete mess right now. A proposal was made (#318) to change the copyright of all files to the two following lines:

> Copyright (c) 2002-2016, Universite catholique de Louvain (UCL), Belgium
> Copyright (c) 2002-2016, OpenJPEG contributors

UCL is there as initiator, host, maintainer, funder, and main contributor of the project. A file would list all OpenJPEG contributors. For now, we did not yet manage to achieve this simplification but concerning the 4 new files you added, would you agree to:

   1. Add the UCL copyright line
   2. Give your consent to the proposal in #318 (add your comment there) so that these new files will not prevent us to move forward on this later on.

* There are plenty of changes in the modified files that are due to space / tab addition / removal. Do you have an easy way to fix this ? Otherwise I would merge your PR in a new branch and fix this from there. Or keep this for when we tackle #128 :-)

Cheers and thanks again.

**Comment by boxerab** [comment]
Hi Antonin,

I will try to remove what I can of the spurious whitespace in the branch. 

Also, I am happy to change the copyright notice on my two files.

Cheers,
Aaron

**Comment by boxerab** [comment]
@detonin  well, I've done what I could to fix the whitespace.  How does it look now?

**Comment by boxerab** [comment]
This PR has been superseded by my decode_region PR. Closing.

**Pull request closed without merging**